common/dbconfig.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright 2021-present Pinterest. All Rights Reserved.
+//
+// @author Premkumar (prem@pinterest.com)
+//
+
+#include "common/dbconfig.h"
+#include "common/segment_utils.h"
+#include <utility>
+
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+
+DECLARE_int32(replicator_replication_mode);
+
+namespace common {
+
+DBConfigManager* DBConfigManager::get() {
+  static DBConfigManager instance;
+  return &instance;
+}
+
+DBConfigManager::DBConfigManager() : dbConfig(new DBConfig()) {
+}
+
+bool DBConfigManager::loadJsonStr(const std::string& content) {
+  Json::Value root;
+  Json::Reader reader;
+  if (!reader.parse(content, root) || !root.isObject()) {
+    LOG(ERROR) << "Could not load config";
+    return false;
+  }
+
+  return loadJsonObject(root);
+}
+
+bool DBConfigManager::loadJsonObject(const Json::Value& root) {
+  try {
+    auto newConfig = parseConfig(root);
+    if (newConfig == nullptr) {
+      return false;
+    }
+
+    LOG(INFO) << "Successfully updated config";
+    std::atomic_exchange_explicit(&dbConfig, newConfig, std::memory_order_release);
+    return true;
+  } catch (std::exception& ex) {
+    LOG(ERROR) << "Exception from parseConfig: " << ex.what();
+    return false;
+  }
+}
+
+std::shared_ptr<const DBConfig> DBConfigManager::parseConfig(const Json::Value& root) {
+  static const std::string ACK_MODE = "ack_mode";
+  static const std::string DATASET = "dataset";
+
+  DBConfig newDBConfig;
+
+  if (root.isMember(DATASET)) {
+    // parse dataset specific config
+    auto jsonDataSetConfigs = root[DATASET];
+
+    for (const auto& dataset : jsonDataSetConfigs.getMemberNames()) {
+      Config config;
+      const auto& jsonDataSetConfig = jsonDataSetConfigs[dataset];
+      bool isSet = false; /* creating this for future multiple assigns */
+
+      if(jsonDataSetConfig.isMember(ACK_MODE)) {
+        config.replication_mode = jsonDataSetConfig[ACK_MODE].asUInt();
+        isSet = true;
+      }
+
+      if (isSet) {
+        newDBConfig.dataSetConfigMap[dataset] = config;
+      }
+    }
+  }
+  return std::make_shared<const DBConfig>(std::move(newDBConfig));
+}
+
+/* conf getter functions */
+const Config* DBConfigManager::getConfig(const std::string& dbName) const {
+  auto dsName = DbNameToSegment(dbName);
+  const auto& dsConfigMap = std::atomic_load_explicit(&dbConfig, std::memory_order_acquire)->dataSetConfigMap;
```

Comment:
C++ questions:
it seems we just need to have read/write locks here (write lock on the write path, read lock on the read path), what's the best practice we do with C++? or why we chose `atomic_load_explicit` and `atomic_exchange_explicit`?

Reply:
I did some search, and find this: https://stackoverflow.com/a/15067071, and https://stackoverflow.com/a/18754519.

Lock is expensive; while, here: all we need is, read and operations after the read are ensured to see the written value. Thus, we need a synchronization between the write/read. 
refer to : https://www.modernescpp.com/index.php/acquire-release-fences

Reply:
looking at https://www.cplusplus.com/reference/atomic/atomic_exchange_explicit/, it takes a `std::atomic<T>*`, but `&dbConfig` isn't? How does this work without a lock

Reply:
The functions are for swapping shared_ptrs atomically ..

Reply:
got it, thanks. https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic